# INFSCI 2725 Assignment 7

Assignmen group 2

Luda Wang luw20@pitt.edu
Kewei Li kel137@pitt.edu
Zheng Liu zhengliu@pitt.edu

In [39]:
#create function can do the parameter tunning and feature selection using different kind of model.

def train(features,label,model_name,para_dict={},feature_importance=False,feature_number=all,print_feature_name=False,feature_model=None,state=0):
    import pandas as pd
    import numpy as np
    import xgboost
    from sklearn.ensemble import AdaBoostClassifier
    from sklearn.ensemble import BaggingClassifier
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.linear_model import LogisticRegression
    from sklearn.model_selection import GridSearchCV
    from sklearn.feature_selection import chi2
    from sklearn.tree import DecisionTreeClassifier
    from sklearn import svm
    
    X=features.values
    
    if(type(label)==pd.core.frame.DataFrame):
        label=label.values.ravel()
        
    if(feature_number!=all):
        
        from sklearn.feature_selection import SelectKBest
    
        fsm=SelectKBest(feature_model,k=feature_number)
        X=fsm.fit_transform(X,label)
        if(print_feature_name==True):
            print("features we used",features.keys()[fsm.get_support(indices=True)])
    
    if model_name=='xgb':
        Model=xgb.XGBClassifier()
        title='XGBoosting'
    elif model_name=='ada':
        Model=AdaBoostClassifier(random_state=state)
        title='Adaptive Boost Classifier'
    elif model_name=='bag':
        Model=BaggingClassifier(random_state=state)
        title='Bagging Classifier'
    elif model_name=='rf':
        Model=RandomForestClassifier(random_state=state)
        title='Random Forest Classifier'
    elif model_name=='lgr':
        Model=LogisticRegression(random_state=state)
        title='Logistic Regression'
    elif model_name=='svc':
        Model=svm.SVC(random_state=state)
        title='Support Vector Classifier'
    elif model_name=='dt':
        Model=DecisionTreeClassifier(random_state=state)
        title='Decision Tree Classifier'
    grid=GridSearchCV(estimator=Model,param_grid=para_dict)
    grid.fit(X,label)
    print("Model name:",title,"\n""The best parameters are",grid.best_params_)
    print("Best Score is",grid.best_score_)

# Iris_data Classification

In [2]:
from sklearn.datasets import load_iris
from sklearn.feature_selection import *
import pandas as pd
import numpy as np
iris_dataset=load_iris() 
print('Keys of the iris_dataset is:', iris_dataset.feature_names)

Keys of the iris_dataset is: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [3]:
feature=pd.DataFrame(iris_dataset.data,columns=iris_dataset.feature_names)

##  SVM classifier 

In [28]:
for i in list(range(1,5)):
    print("number of features",i)
    train(feature,iris_dataset.target,'svc',[{'kernel': ['rbf'], 
                                          'gamma': np.arange(0.0005,0.0015,0.0001),
                                          'C': list(range(630,660,1))},
                                         {'kernel': ['linear'], 
                                          'C': list(range(1,1000,100))},
                                         {'kernel':['poly'],
                                          'C':list(range(1,1000,100)),
                                          'degree':[2,3,4]}],
          feature_model=chi2,
          feature_number=i,
          print_feature_name=True)

number of features 1
features we used Index(['petal length (cm)'], dtype='object')
Model name: Support Vector Classifier 
The best parameters are {'C': 101, 'degree': 4, 'kernel': 'poly'}
Best Score is 0.96
number of features 2
features we used Index(['petal length (cm)', 'petal width (cm)'], dtype='object')
Model name: Support Vector Classifier 
The best parameters are {'C': 630, 'gamma': 0.0007000000000000001, 'kernel': 'rbf'}
Best Score is 0.9733333333333334
number of features 3
features we used Index(['sepal length (cm)', 'petal length (cm)', 'petal width (cm)'], dtype='object')
Model name: Support Vector Classifier 
The best parameters are {'C': 630, 'gamma': 0.0005, 'kernel': 'rbf'}
Best Score is 0.98
number of features 4
features we used Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')
Model name: Support Vector Classifier 
The best parameters are {'C': 636, 'gamma': 0.0011000000000000003, 'kernel': 'rbf'}
Bes

Comment: 

we use Chi2 feature selection model and GridSearch method to do the feature selection, feature expansion(kernal) and parameter tunning. Within GridSearch, 3-fold cross validation is used automatically. The best score we get is using kernel 'rbf' and all 4 features.

## Logistic Regression

In [34]:
for i in list(range(1,5)):
    print("number of features",i)
    train(feature,iris_dataset.target,'lgr',[{'penalty':['l1'],
                                          'C':list(range(1,100))},
                                         {'penalty':['l2'],
                                          'C':list(range(1,100))}],
          feature_number=i,feature_model=chi2)

number of features 1
Model name: Logistic Regression 
The best parameters are {'C': 3, 'penalty': 'l1'}
Best Score is 0.9533333333333334
number of features 2
Model name: Logistic Regression 
The best parameters are {'C': 5, 'penalty': 'l1'}
Best Score is 0.9666666666666667
number of features 3
Model name: Logistic Regression 
The best parameters are {'C': 5, 'penalty': 'l1'}
Best Score is 0.9666666666666667
number of features 4
Model name: Logistic Regression 
The best parameters are {'C': 61, 'penalty': 'l2'}
Best Score is 0.9733333333333334


Comment:

All 4 features should be used to get the largest score when using Logistic Regression.

In [38]:
for i in list(range(1,5)):
    print("number of features",i)
    train(feature,iris_dataset.target,'rf',{'max_features':['auto','log2',None],
                                        'n_estimators':list(range(1,100,10)),
                                        'min_samples_leaf':list(range(1,10,1))},
      feature_model=chi2,
      feature_number=i)

number of features 1
Model name: Random Forest Classifier 
The best parameters are {'max_features': 'auto', 'min_samples_leaf': 2, 'n_estimators': 51}
Best Score is 0.9533333333333334
number of features 2
Model name: Random Forest Classifier 
The best parameters are {'max_features': 'auto', 'min_samples_leaf': 2, 'n_estimators': 31}
Best Score is 0.9733333333333334
number of features 3
Model name: Random Forest Classifier 
The best parameters are {'max_features': 'auto', 'min_samples_leaf': 2, 'n_estimators': 31}
Best Score is 0.9733333333333334
number of features 4
Model name: Random Forest Classifier 
The best parameters are {'max_features': 'auto', 'min_samples_leaf': 2, 'n_estimators': 31}
Best Score is 0.9733333333333334


Comment:

As we can see, the score is not changed when adding more features after 2. Therefore, to avoid overfitting, we can use 2 features when apply Random Forest Classifier. Two features we selected from Chi2 score are 'petal length (cm)' and 'petal width (cm)'.

## Adaptive Boosting Model

In [40]:
for i in list(range(1,5)):
    print("number of features",i)
    train(feature,iris_dataset.target,'ada',{'n_estimators':list(range(5,15,1)),
                                         'learning_rate':np.arange(0.5,1.5,0.1)},
          feature_model=chi2,
      feature_number=i)

number of features 1
Model name: Adaptive Boost Classifier 
The best parameters are {'learning_rate': 1.0999999999999999, 'n_estimators': 13}
Best Score is 0.9533333333333334
number of features 2
Model name: Adaptive Boost Classifier 
The best parameters are {'learning_rate': 1.1999999999999997, 'n_estimators': 8}
Best Score is 0.9733333333333334
number of features 3
Model name: Adaptive Boost Classifier 
The best parameters are {'learning_rate': 1.1999999999999997, 'n_estimators': 5}
Best Score is 0.9733333333333334
number of features 4
Model name: Adaptive Boost Classifier 
The best parameters are {'learning_rate': 1.1999999999999997, 'n_estimators': 8}
Best Score is 0.9733333333333334


Comment:

Two features we need to use. same as above.

## Decision Tree Classifier

In [43]:
for i in list(range(1,5)):
    print("number of features",i)
    train(feature,iris_dataset.target,'dt',{'max_depth':list(range(1,15,1)),
                                       'min_samples_leaf':list(range(1,10,1)),
                                       'min_samples_split':list(range(2,10,1))},
          feature_model=chi2,
          feature_number=i)

number of features 1
Model name: Decision Tree Classifier 
The best parameters are {'max_depth': 2, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Score is 0.94
number of features 2
Model name: Decision Tree Classifier 
The best parameters are {'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Score is 0.96
number of features 3
Model name: Decision Tree Classifier 
The best parameters are {'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Score is 0.96
number of features 4
Model name: Decision Tree Classifier 
The best parameters are {'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Score is 0.9666666666666667


Comment:

4 features should be used to use Decision Tree Classifier.

The highest score we get is from SVM classifier with all 4 features and kernal 'rbf' . The best score is 0.993.

# Congressional Voting Records Classification 

In [52]:
votes=pd.read_table('house-votes-84.txt',index_col=None)

target_vote=pd.concat([votes[['Party']]])
features_vote=votes.drop(['Party'], axis=1)

features_vote.replace(to_replace='y',value=1,inplace=True)
features_vote.replace(to_replace='w',value=2,inplace=True)
features_vote.replace(to_replace='n',value=0,inplace=True)



## SVM Classifier

In [65]:
for i in list(range(1,17)):
    print("number of features",i)
    train(features_vote,target_vote,'svc',[{'kernel': ['rbf'], 
                                        'gamma': [1e-3, 1e-4],
                                        'C': list(range(1,100,10))},
                                       {'kernel': ['linear'], 
                                        'C': list(range(1,1000,100))}],feature_number=i,feature_model=chi2
)

number of features 1
Model name: Support Vector Classifier 
The best parameters are {'C': 11, 'gamma': 0.001, 'kernel': 'rbf'}
Best Score is 0.9448275862068966
number of features 2
Model name: Support Vector Classifier 
The best parameters are {'C': 21, 'gamma': 0.001, 'kernel': 'rbf'}
Best Score is 0.9448275862068966
number of features 3
Model name: Support Vector Classifier 
The best parameters are {'C': 11, 'gamma': 0.001, 'kernel': 'rbf'}
Best Score is 0.9471264367816092
number of features 4
Model name: Support Vector Classifier 
The best parameters are {'C': 31, 'gamma': 0.001, 'kernel': 'rbf'}
Best Score is 0.9471264367816092
number of features 5
Model name: Support Vector Classifier 
The best parameters are {'C': 51, 'gamma': 0.001, 'kernel': 'rbf'}
Best Score is 0.9448275862068966
number of features 6
Model name: Support Vector Classifier 
The best parameters are {'C': 41, 'gamma': 0.001, 'kernel': 'rbf'}
Best Score is 0.9448275862068966
number of features 7
Model name: Support

comment:
    
The best number of features is 3,kernel is 'rbf',and the features we choose are 'adoption_of_the_budget_resolution', 'physician_fee_freeze' and 'education_spending'.

## Logistic Regression

In [60]:
for i in list(range(1,17)):
    print("number of features",i)
    train(features_vote,target_vote,'lgr',[{'penalty':['l1'],
                                        'C':list(range(1,100))},
                                       {'penalty':['l2'],
                                        'C':list(range(1,100))}],feature_model=chi2,feature_number=i)

number of features 1
Model name: Logistic Regression 
The best parameters are {'C': 1, 'penalty': 'l1'}
Best Score is 0.9448275862068966
number of features 2
Model name: Logistic Regression 
The best parameters are {'C': 1, 'penalty': 'l1'}
Best Score is 0.9448275862068966
number of features 3
Model name: Logistic Regression 
The best parameters are {'C': 1, 'penalty': 'l1'}
Best Score is 0.9333333333333333
number of features 4
Model name: Logistic Regression 
The best parameters are {'C': 1, 'penalty': 'l1'}
Best Score is 0.9287356321839081
number of features 5
Model name: Logistic Regression 
The best parameters are {'C': 1, 'penalty': 'l1'}
Best Score is 0.9310344827586207
number of features 6
Model name: Logistic Regression 
The best parameters are {'C': 9, 'penalty': 'l2'}
Best Score is 0.9333333333333333
number of features 7
Model name: Logistic Regression 
The best parameters are {'C': 1, 'penalty': 'l1'}
Best Score is 0.9310344827586207
number of features 8
Model name: Logistic

Comment:

Logistic Regression uses 15 features to get the best score. The best score is 0.9494

## Random Forest Classifier

In [61]:
for i in list(range(1,17)):
    print("number of features",i)
    train(features_vote,target_vote,'rf',{'max_features':['auto','log2',None],
                                      'n_estimators':list(range(15,25,1)),
                                      'min_samples_leaf':list(range(1,10,1))},feature_number=i,feature_model=chi2)

number of features 1
Model name: Random Forest Classifier 
The best parameters are {'max_features': 'auto', 'min_samples_leaf': 1, 'n_estimators': 15}
Best Score is 0.9563218390804598
number of features 2
Model name: Random Forest Classifier 
The best parameters are {'max_features': 'auto', 'min_samples_leaf': 2, 'n_estimators': 15}
Best Score is 0.9586206896551724
number of features 3
Model name: Random Forest Classifier 
The best parameters are {'max_features': 'auto', 'min_samples_leaf': 1, 'n_estimators': 16}
Best Score is 0.9563218390804598
number of features 4
Model name: Random Forest Classifier 
The best parameters are {'max_features': 'auto', 'min_samples_leaf': 5, 'n_estimators': 20}
Best Score is 0.9563218390804598
number of features 5
Model name: Random Forest Classifier 
The best parameters are {'max_features': None, 'min_samples_leaf': 4, 'n_estimators': 17}
Best Score is 0.9586206896551724
number of features 6
Model name: Random Forest Classifier 
The best parameters are

Comment:

15 features are selected applying Random Forest Classier. The best score is 0.9747.

## Adaptive Boosting Model

In [64]:
for i in list(range(1,17)):
    print("number of features",i)
    train(features_vote,target_vote,'ada',{'n_estimators':list(range(1,100,10)),
                                       'learning_rate':np.arange(0.001,0.012,0.01)},
      feature_model=chi2,
      feature_number=i)

number of features 1
Model name: Adaptive Boost Classifier 
The best parameters are {'learning_rate': 0.011, 'n_estimators': 51}
Best Score is 0.9471264367816092
number of features 2
Model name: Adaptive Boost Classifier 
The best parameters are {'learning_rate': 0.011, 'n_estimators': 71}
Best Score is 0.9471264367816092
number of features 3
Model name: Adaptive Boost Classifier 
The best parameters are {'learning_rate': 0.001, 'n_estimators': 1}
Best Score is 0.9448275862068966
number of features 4
Model name: Adaptive Boost Classifier 
The best parameters are {'learning_rate': 0.001, 'n_estimators': 1}
Best Score is 0.9448275862068966
number of features 5
Model name: Adaptive Boost Classifier 
The best parameters are {'learning_rate': 0.001, 'n_estimators': 1}
Best Score is 0.9448275862068966
number of features 6
Model name: Adaptive Boost Classifier 
The best parameters are {'learning_rate': 0.001, 'n_estimators': 1}
Best Score is 0.9448275862068966
number of features 7
Model name:

Comment:
    
3 features are selected applying adaptive boosting model. The best score is 0.9448

## Decision Tree Classifier

In [63]:
for i in list(range(1,17)):
    print("number of features",i)
    train(features_vote,target_vote,'dt',{'max_depth':list(range(1,15,1)),
                                       'min_samples_leaf':list(range(1,10,1)),
                                       'min_samples_split':list(range(2,10,1))},feature_model=chi2,feature_number=i)

number of features 1
Model name: Decision Tree Classifier 
The best parameters are {'max_depth': 2, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Score is 0.9563218390804598
number of features 2
Model name: Decision Tree Classifier 
The best parameters are {'max_depth': 2, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Score is 0.9563218390804598
number of features 3
Model name: Decision Tree Classifier 
The best parameters are {'max_depth': 3, 'min_samples_leaf': 2, 'min_samples_split': 2}
Best Score is 0.9563218390804598
number of features 4
Model name: Decision Tree Classifier 
The best parameters are {'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Score is 0.9563218390804598
number of features 5
Model name: Decision Tree Classifier 
The best parameters are {'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Score is 0.9563218390804598
number of features 6
Model name: Decision Tree Classifier 
The best parameters are {'max_depth': 3, 

Comment:

12 features are selected applying Decision Tree Classier. The best Score is 0.9724

The best score we get is from model Decision Tree Classier. The best score is 0.9724